runtime/src/main/java/org/corfudb/runtime/view/workflows/WorkflowRequest.java

```diff
@@ -0,0 +1,140 @@
+     * are on unresponsive nodes and the affected endpoint by the workflow
+     * will not be selected.
+     *
+     * @param layout the layout that is used to select an orchestrator
+     * @return a management client that is connected to the selected
+     * orchestrator
+     */
+    protected ManagementClient getOrchestrator(Layout layout) {
+        List<String> activeLayoutServers = layout.getLayoutServers().stream()
+                .filter(s -> !layout.getUnresponsiveServers().contains(s)
+                        && !s.equals(nodeForWorkflow))
+                .collect(Collectors.toList());
+
+        if (activeLayoutServers.isEmpty()) {
+            throw new WorkflowException("getOrchestrator: no available orchestrators " + layout);
+        }
+
+        return runtime.getRouter(activeLayoutServers.get(0)).getClient(ManagementClient.class);
+    }
+
+    /**
+     * Infer the completion of the request by inspecting the layout
+     *
+     * @param layout the layout to inspect
+     * @return true if the operation is reflected in the layout, otherwise
+     * return false
+     */
+    protected abstract boolean verifyRequest(Layout layout);
+
+    /**
+     * @param workflow   the workflow id to poll
+     * @param client     a client that is connected to the orchestrator that is
+     *                   running the workflow
+     * @param timeout    the total time to wait for the workflow to complete
+     * @param pollPeriod the poll period to query the completion of the workflow
+     * @throws TimeoutException if the workflow doesn't complete withint the timout
+     *                          period
+     */
+    private void waitForWorkflow(@Nonnull UUID workflow, @Nonnull ManagementClient client,
+                                 @Nonnull Duration timeout, @Nonnull Duration pollPeriod)
+            throws TimeoutException {
+        long tries = timeout.getSeconds() / pollPeriod.getSeconds();
+        for (long x = 0; x < tries; x++) {
+            if (!client.queryRequest(workflow).isActive()) {
+                return;
+            }
+            Sleep.sleepUninterruptibly(pollPeriod);
+            log.info("waitForWorkflow: waiting for {} on attempt {}", workflow, x);
+        }
+        throw new TimeoutException();
+    }
+
+    /**
+     * Starts executing the workflow request.
+     *
+     * This method will succeed only if the workflow executed successfully, otherwise
+     * it can throw a WorkflowResultUnknownException when the timeouts are exhauseted
+     * and the expected side effect cannot be verified.
+     *
+     * @throws WorkflowResultUnknownException when the workflow result cannot be
+     * verified.
+     */
+    public void invoke() {
+        for (int x = 0; x < retry; x++) {
+            try {
+                runtime.invalidateLayout();
+                Layout requestLayout = new Layout(runtime.getLayoutView().getLayout());
+                UUID workflowId = sendRequest(requestLayout);
```

Comment:
nit: this got me confused. sendRequest uses the same orchestrator as you obtain in the next line. The connection is not very apparent in the code.